File: c2safeRust_examples/bst.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int key;
    char* value;
    struct Node* left;
    struct Node* right;
} node_t;

void printAll(node_t* root, int level) {
    if (root->left) {
        printAll(root->left, level+1);
    }
    for (int i = 0; i < level; i++) {
        printf("    ");
    }
    printf("%d:%s\n", root->key, root->value);
    if (root->right) {
        printAll(root->right, level+1);
    }
}

void freeAll(node_t* root) {
    if (root->left) {
        freeAll(root->left);
    }
    if (root->right) {
        freeAll(root->right);
    }
    root->key = 0;
    free(root);
}

void insert(node_t* root, int k, char* v) {
    if (k < root->key) {
        if (root->left) {
            insert(root->left, k, v);
        } else {
            root->left = malloc(sizeof(node_t));
            root->left->key = k;
            root->left->value = v;
            root->left->left = NULL;
            root->left->right = NULL;
        }
    } else if (k > root->key) {
        if (root->right) {
            insert(root->right, k, v);
        } else {
            root->right = malloc(sizeof(node_t));
            root->right->key = k;
            root->right->value = v;
            root->right->left = NULL;
            root->right->right = NULL;
        }
    } else {
        root->value = v;
    }
}

char* search(node_t* root, int k) {
    if (k < root->key) {
        if (root->left) {
            return search(root->left, k);
        } else {
            return NULL;
        }
    } else if (k > root->key) {
        if (root->right) {
            return search(root->right, k);
        } else {
            return NULL;
        }
    } else {
        return root->value;
    }
}

int main(int argc, char* argv[]) {
    node_t* root = malloc(sizeof(node_t));
    root->key = 5;
    root->value = "five";
    root->left = NULL;
    root->right = NULL;
    char* returnholder;
    returnholder = search(root, 5);
    printf("5: %s\n", returnholder);
    returnholder = search(root, 3);
    printf("3: %s\n", returnholder);
    insert(root, 3, "three");
    returnholder = search(root, 3);
    printf("3: %s\n", returnholder);
    returnholder = search(root, 7);
    printf("7: %s\n", returnholder);
    insert(root, 7, "seven");
    returnholder = search(root, 7);
    printf("7: %s\n", returnholder);
    returnholder = search(root, 4);
    printf("4: %s\n", returnholder);
    insert(root, 4, "four");
    returnholder = search(root, 4);
    printf("4: %s\n", returnholder);
    returnholder = search(root, 2);
    printf("2: %s\n", returnholder);
    insert(root, 2, "two");
    returnholder = search(root, 2);
    printf("2: %s\n", returnholder);
    returnholder = search(root, 6);
    printf("6: %s\n", returnholder);
    insert(root, 6, "six");
    returnholder = search(root, 6);
    printf("6: %s\n", returnholder);
    returnholder = search(root, 8);
    printf("8: %s\n", returnholder);
    insert(root, 8, "eight");
    returnholder = search(root, 8);
    printf("8: %s\n", returnholder);
    printAll(root, 0);
    freeAll(root);
}

File: source/pyggi/sample/bst_rust/bst.rs
#[derive(Debug)]
struct Node {
    key: i32,
    value: String,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
    parent_ptr: *mut Node, // raw pointer to parent node
}

// static mutable counter of nodes (example of global unsafe variable)
static mut NODE_COUNT: i32 = 0;

impl Node {
    fn new(key: i32, value: String) -> Self {
        unsafe { NODE_COUNT += 1; } // increment global counter
        Node {
            key,
            value,
            left: None,
            right: None,
            parent_ptr: std::ptr::null_mut(), // initially null
        }
    }

    fn insert(&mut self, key: i32, value: String) {
        if key < self.key {
            if let Some(ref mut left) = self.left {
                // update parent_ptr for child
                unsafe { left.parent_ptr = self; }
                left.insert(key, value);
            } else {
                let mut new_node = Box::new(Node::new(key, value));
                unsafe { new_node.parent_ptr = self; } // set raw pointer to parent
                self.left = Some(new_node);
            }
        } else if key > self.key {
            if let Some(ref mut right) = self.right {
                unsafe { right.parent_ptr = self; }
                right.insert(key, value);
            } else {
                let mut new_node = Box::new(Node::new(key, value));
                unsafe { new_node.parent_ptr = self; }
                self.right = Some(new_node);
            }
        } else {
            self.value = value;
        }
    }

    fn search(&mut self, key: i32) -> Option<&String> {
        if key < self.key {
            if let Some(ref mut left) = self.left {
                left.search(key)
            } else {
                None
            }
        } else if key > self.key {
            if let Some(ref mut right) = self.right {
                right.search(key)
            } else {
                None
            }
        } else {
            // unsafe raw pointer dereference example
            unsafe {
                if !self.parent_ptr.is_null() {
                    println!("Parent key (via raw pointer): {}", (*self.parent_ptr).key);
                }
            }
            Some(&self.value)
        }
    }
}

fn main() {
    let mut root = Node::new(5, String::from("five"));

    println!("{:?}", root.search(5));
    root.insert(3, String::from("three"));
    println!("{:?}", root.search(3));
    root.insert(7, String::from("seven"));
    println!("{:?}", root.search(7));
    root.insert(4, String::from("four"));
    println!("{:?}", root.search(4));
    root.insert(2, String::from("two"));
    println!("{:?}", root.search(2));
    root.insert(6, String::from("six"));
    println!("{:?}", root.search(6));
    root.insert(8, String::from("eight"));
    println!("{:?}", root.search(8));

    unsafe {
        println!("Total nodes (via static mutable global): {}", NODE_COUNT);
    }
}
